### PR TITLE
#9377: Merge binary ops from tt_eager into TTNN

### LINKED ISSUE #9377: Move eltwise binary to ttnn

### PR DESCRIPTION
**Pull request by ayerofieiev-tt**
### Ticket
#9377

### Problem description
We want to smoke out issues related to the attempt to merge binary op from tt_eager to ttnn.
If succesfull, we build more examples of how to make the transition.

### What changed
Eltwise `binary ops` consolidated in `ttnn`.
From 😫
```
tt_eager
└── tt_dnn
    └── op_library
        ├── ...
        └── eltwise_binary
            ├── kernels
            │   ├── compute
            │   │   └── eltwise_binary.cpp
            │   └── dataflow
            │       └── reader_binary_interleaved_start_id.cpp
            ├── multi_core
            │   └── eltwise_binary_op_multi_core.cpp
            ├── eltwise_binary_op.cpp
            └── eltwise_binary_op.hpp
ttnn
└── cpp
    ├── pybind11
    │   └── operations
    │       ├── ...
    │       └── binary.hpp
    └── ttnn
        ├── op_library
        │   └── ...
        │   └── binary
        │       ├── binary_op.hpp
        │       └── binary_op.cpp
        └── operations
            ├── ...
            └── binary.hpp
```
to 😎
```
ttnn
└── cpp
    └── ttnn
        └── operations
            └── ...
            └── eltwise
                └── binary
                    ├── device
                    │   ├── kernels
                    │   │   ├── compute
                    │   │   │   └── eltwise_binary.cpp
                    │   │   └── dataflow
                    │   │       └── reader_binary_interleaved_start_id.cpp
                    │   ├── binary_op.cpp
                    │   ├── binary_op.hpp
                    │   └── binary_program_factory.hpp
                    ├── binary.hpp
                    └── binary_pybind.hpp
```

### Findings
- Many parts of the code depend on eltwise binary ops, specially composite_ops
- Such change forces some ttnn includes in tt_dnn, which gets/near gets us to circular dependency

### Note on the PR size
Apologies for the >130 files changed.
For this work I now prefer to not redo it and take one person to patiently review changes made here.
From this experience I learn a better way to do it going forward.

### Checklist
- [x] Post commit CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/9523682917/job/26256854507)
- [x] [post-commit] models tests [passes](https://github.com/tenstorrent/tt-metal/actions/runs/9505032784)
- [x] Device perf regressions and output report [passes](https://github.com/tenstorrent/tt-metal/actions/runs/9505239528/job/26199677516) 
- [x] Model perf regressions and output report [passes](https://github.com/tenstorrent/tt-metal/actions/runs/9505238567)
- [x] Nightly fast dispatch model regression CI has same result as main [latest](https://github.com/tenstorrent/tt-metal/actions/runs/9504975304)
- [ ] T300 frequent tests passes (can't run, broken on main)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt_eager/tt_dnn/op_library/loss/loss_op.cpp
Comment (ayerofieiev-tt) [inline tt_eager/tt_dnn/op_library/loss/loss_op.cpp:1]: @arakhmati lets switch from vector<strings> to accept a proper structure? any concerns?

**Comment by cfjchu** [comment]
+1 on the PR description..

> Such change forces some ttnn includes in tt_dnn, which gets/near gets us to circular dependency

What is the plan here? Also, provided this is successful, what's the transition plan for the other ops?

**Comment by ayerofieiev-tt** [comment]
@cfjchu , here is the plan
https://github.com/tenstorrent/tt-metal/issues/9322

It is still in progress as with the PR I try to find what stands in out way and then Borys, Naif, Nitika, MCW folks and other ops leads will start moving their specific ops. I am most worried about complexity around python due to it be interpreted..

File: CMakeLists.txt [hunk 1]
```diff
@@ -150,12 +150,12 @@ else()
 endif()
 
 # Can't use the `REUSE_FROM` option bc tt_lib and ttnn have different build flags :(
-add_library(pch_pybinds INTERFACE)
-target_precompile_headers(pch_pybinds INTERFACE
-    ${PROJECT_SOURCE_DIR}/tt_metal/third_party/pybind11/include/pybind11/operators.h
-    ${PROJECT_SOURCE_DIR}/tt_metal/third_party/pybind11/include/pybind11/pybind11.h
-    ${PROJECT_SOURCE_DIR}/tt_metal/third_party/pybind11/include/pybind11/stl.h
-)
+# add_library(pch_pybinds INTERFACE)
```
Comment (ayerofieiev-tt) [inline CMakeLists.txt:153]: PCH to be disabled for now. Due to addition of a link between tt_eager and ttnn_lib something causes pch to get updated in the middle of the build. Will turn PCH back on when tt_eager and ttnn are merged.

File: models/demos/metal_BERT_large_11/tt/embeddings.py [hunk 2]
```diff
@@ -153,7 +154,7 @@ def __call__(
         token_type_ids.deallocate()
 
         if self.position_embedding_type == "absolute":
-            inputs_plus_token_type_embeddings_tt_tensor = ttl.tensor.add(
+            inputs_plus_token_type_embeddings_tt_tensor = ttnn.add(
```
Comment (ayerofieiev-tt) [inline models/demos/metal_BERT_large_11/tt/embeddings.py:157]: @arakhmati this PR could be way smaller if I was smarter and created an alias. I would use some help here.
W/o an alias, one has to update all tests and all models, which is extremely involved.
Instead, I propose to carry an alias until we are close to the completion of the migration. Wdyt?

Most of the changes one sees in this PR is a switch from `ttl.tensor.add/mul/sub...` or `tt_lib.add..` to ttnn analog.

**Comment by ayerofieiev-tt** [comment]
This PR shows the worst approach possible.
We discussed with Akhmed an alternative, which can help to split the future scope into more manageable and reviewable pieces.
TTNN Migration

So in case of complex areas like Unary/ Binary one would first op-by-op migrate existing code and tests to ttnn and only then remove tt_lib/ttl.tensor bindings and C++ code. This allows to run migration in parallel, with small reviewable PRs.

File: tests/tt_eager/profiling/ops_for_profiling.py [hunk 3]
```diff
@@ -421,15 +421,15 @@ def angle_bw(x, y):
 
 all_binary_ops = [
     {
-        "op": tt_lib.tensor.add,
+        "op": ttnn.add,
         "name": "tt_lib.tensor.add",
```
Comment (ayerofieiev-tt) [inline tests/tt_eager/profiling/ops_for_profiling.py:425]: Need to change?

File: tests/ttnn/unit_tests/operations/test_add.py [hunk 4]
```diff
@@ -209,7 +209,9 @@ def test_add_dram_and_l1_tensor(device, shape_a, shape_b):
 
 
 @pytest.mark.parametrize("shape", [(1, 1, 32, 32)])
-@pytest.mark.parametrize("activations", [None, ["relu"]])
+@pytest.mark.parametrize(
+    "activations", [None, [ttl.tensor.FusibleActivationWithParam(ttl.tensor.FusibleActivation.RELU)]]
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_add.py:213]: happy about this change

File: ttnn/cpp/ttnn/operations/conv2d.hpp [hunk 5]
```diff
@@ -82,8 +82,6 @@ struct Conv2dConfig {
     }
 };
 
-extern const std::array<ttnn::TensorSchema, 3> input_schemas;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv2d.hpp:85]: This caused ODR violation

**Review by arakhmati — CHANGES_REQUESTED** [review]

**Review by arakhmati — APPROVED** [review]

File: ttnn/ttnn/operations/binary.py [hunk 6]
```diff
@@ -246,6 +246,10 @@ def _golden_function(input_tensor_a, input_tensor_b, *args, **kwargs):
 
 logaddexp2 = ttnn.register_operation(golden_function=_golden_function)(ttnn._ttnn.operations.binary.logaddexp2)
 
+divide = ttnn.register_operation()(ttnn._ttnn.operations.binary.divide)
```
Comment (arakhmati) [inline ttnn/ttnn/operations/binary.py:249]: Would be nice to have the golden function implemented for `divide` and `bias_gelu`. Can be done on a future PR

Comment (ayerofieiev-tt) [reply ttnn/ttnn/operations/binary.py]: added

File: CMakeLists.txt (hunk 1 above)
Comment (ayerofieiev-tt) [reply CMakeLists.txt:151]: Linking issue here
https://github.com/tenstorrent/tt-metal/issues/9457

**Review by cfjchu — APPROVED** [review]

**Review by mywoodstock — APPROVED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (tt-rkim) [reply CMakeLists.txt:151]: Might wanna leave a comment within the `CMakeLists.txt` with a link to that issue so it doesn't get lost

File: tt_eager/CMakeLists.txt [hunk 7]
```diff
@@ -11,12 +11,13 @@ set(TT_EAGER_OBJECTS
 
 # Defaults to a shared lib, static lib for production python wheel
 add_library(tt_eager ${TT_EAGER_OBJECTS})
-target_link_libraries(tt_eager PUBLIC compiler_flags linker_flags tt_metal)     # linker_flags = -rdynamic if tracy enabled
+target_link_libraries(tt_eager PUBLIC compiler_flags linker_flags tt_metal ttnn_lib)     # linker_flags = -rdynamic if tracy enabled
```
Comment (tt-rkim) [inline tt_eager/CMakeLists.txt:14]: Damn... really hope all of `tt_eager` can go bye bye soon

File: CMakeLists.txt (hunk 1 above)
Comment (ayerofieiev-tt) [reply CMakeLists.txt:151]: Already opened a PR fixing this

**Pull request merged**